src/liballoc/collections/btree/map.rs

```diff
@@ -525,7 +569,39 @@ impl<K: Ord, V> BTreeMap<K, V> {
         }
     }
 
-    /// Clears the map, removing all values.
+    /// Clears the map, returning all key-value pairs as an iterator.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// #![feature(btree_drain_retain)]
+    /// use std::collections::BTreeMap;
+    ///
+    /// let mut a = BTreeMap::new();
+    /// a.insert(1, "a");
+    /// a.insert(2, "b");
+    ///
+    /// for (k, v) in a.drain().take(1) {
+    ///     assert!(k == 1 || k == 2);
+    ///     assert!(v == "a" || v == "b");
+    /// }
+    ///
+    /// assert!(a.is_empty());
+    /// ```
+    #[unstable(feature = "btree_drain_retain", issue = "42849")]
+    pub fn drain(&mut self) -> IntoIter<K, V> {
```

Comment:
This should return a new `Drain` type, just like `HashMap::drain` does, even if that's still just a newtype wrapping `IntoIter`.

Reply:
I would expect `BTreeSet::drain` to remove a range of keys, much like `Vec::drain` removes a range of indices. There are ways to implement this that are [more efficient than repeated calls to `remove](http://www.cs.hut.fi/u/tlilja/thesis/masters.pdf), but they are pretty complex. I think it would be a mistake to commit to this signature for `drain`, since range deletion is quite useful.

Reply:
I had the same expectation about drain and asked in #42849. I guess the signature should be discussed first in an RFC for drain specifically?